src/orbit_zoom_camera.rs

```diff
@@ -0,0 +1,191 @@
+//! 
+//! A 3dsMax / Blender style camera that orbits about a target position
+//!
+
+use event::GenericEvent;
+use std::ops::Mul;
+use cgmath::*;
+use { input, Camera };
+
+use input::Button::{Keyboard, Mouse};
+use input::keyboard::Key;
+use input::mouse::MouseButton;
+
+
+bitflags!(flags Keys: u8 {
+    const ZOOM  = 0b00000001,
+    const PAN   = 0b00000010,
+    const ORBIT = 0b00000100,
+});
+
+///
+/// Specifies key bindings and speed modifiers for OrbitZoomCamera
+///
+pub struct OrbitZoomCameraSettings<T=f32> {
+
+    /// Which button to press to orbit with mouse
+    pub orbit_button: input::Button,
+
+    /// Which button to press to zoom with mouse
+    pub zoom_button: input::Button,
+
+    /// Which button to press to pan with mouse
+    pub pan_button: input::Button,
+
+    /// Modifier for orbiting speed (arbitrary unit)
+    pub orbit_speed: T,
+
+    /// Modifier for panning speed (arbitrary unit)
+    pub pan_speed: T,
+
+    /// Modifier for zoom speed (arbitrary unit)
+    pub zoom_speed: T,
+}
+
+impl OrbitZoomCameraSettings {
+
+    ///
+    /// Clicking and dragging OR two-finger scrolling will orbit camera,
+    /// with LShift as pan modifer and LCtrl as zoom modifier
+    ///
+    pub fn default() -> OrbitZoomCameraSettings {
+        OrbitZoomCameraSettings {
+            orbit_button : Mouse(MouseButton::Left),
+            zoom_button : Keyboard(Key::LCtrl),
+            pan_button : Keyboard(Key::LShift),
+            orbit_speed: 0.05,
+            pan_speed: 0.1,
+            zoom_speed: 0.1,
+        }
+    }
+
+}
+
+///
+/// A 3dsMax / Blender-style camera that orbits around a target point
+/// 
+pub struct OrbitZoomCamera<T=f32> {
+
+    /// origin of camera rotation
+    pub target: Vector3<T>,
+
+    /// Rotation of camera relative to world-space positive z-axis
+    pub rotation: Quaternion<T>,
+
+    /// Pitch up/down from target
+    pub pitch: T,
+    
+    /// Yaw left/right from target
+    pub yaw: T,
+
+    /// camera distance from target
+    pub distance: T,
+
+    /// Settings for the camera
+    pub settings: OrbitZoomCameraSettings<T>,
+
+    /// Current keys that are pressed
+    keys: Keys,
+}
+
+
+impl OrbitZoomCamera {
+
+    ///
+    /// Create a new OrbitZoomCamera targeting the given coordinates
+    /// 
+    pub fn new(target: [f32; 3], settings: OrbitZoomCameraSettings) -> OrbitZoomCamera {
```

Comment:
Being able to change between `f32` and `f64` is nice, but this can easily be fixed later.